components/learn/modules/ROOT/pages/on-upgrades.adoc

```diff
@@ -1,8 +1,281 @@
+
+NOTE: In future versions of the OpenZeppelin SDK, the CLI will take care of automatically converting constructors into initializers, so you do not need to worry about this.
+
+To learn more about this and other caveats when writing upgradeable contracts, check out our xref:sdk::writing-contracts.adoc[Writing Upgradeable Contracts] guide.
+
+=== Upgrading
+
+Due to technical limitations, when you upgrade a contract to a new version, you cannot change the **storage layout** of that contract.
+
+This means that, if you have already declared a state variable in your contract, you cannot remove it, change its type, or declare another variable before it. In our `Box` example, it means that we can only add new state variables _after_ `value`.
+
+```solidity
+// contracts/Box.sol
+contract Box {
+    uint256 private value;
+
+    // We can safely add a new variable after the ones we had declared
+    address private owner; 
+    
+    // ...
+}
+```
+
+Fortunately, this limitation only affects state variables. You can change the contract's functions and events as you wish.
+
+NOTE: If you accidentally mess up with your contract's storage layout, the CLI will warn you when you try to upgrade.
+
+To learn more about this limitation, head over to the xref:sdk::writing-contracts.adoc#modifying-your-contracts[Modifying Your Contracts] guide.
+
+[[upgrading-contracts-in-js]]
+== Upgrading Contracts Programmatically
+
+If you want to create and upgrade contracts from your JavaScript code instead of via the command line, you can use the `@openzeppelin/upgrades.js` library instead of the CLI.
+
+NOTE: The CLI does not just manage contract upgrades, but also compilation, interaction, and source code verification. The `upgrade.js` library only takes care of creating and upgrading. The library also does not keep track of the contracts you have already deployed, nor runs any initializer or storage layout validations, as the CLI does. Nevertheless, these capabilities may be added to `upgrade.js` in the near future.
+
+Your first step will be to install the library in your project, and you will also probably want to install `web3`:
+
+```bash
+$ npm install @openzeppelin/upgrades.js web3
+```
+
+As in our previous guide on programmatically interacting with contracts, we will start with some boilerplate code to initialize a provider, as well as the upgrades library.
+
+```js
+const Web3 = require('web3');
+const Upgrades = require('@openzeppelin/upgrades')
+
+async function main() {
+  // Set up web3 object, connected to the local development network, initialize the Upgrades library
+  const web3 = new Web3('http://localhost:8545');
+  Upgrades.ZWeb3.initialize(web3.currentProvider)
+}
+
+main();
+```
+
+NOTE: You can check out a full version of the code in this section in the https://github.com/OpenZeppelin/openzeppelin-sdk/tree/master/examples/upgrades-library[upgrades-library example of the SDK repository].
```

Comment:
This link leads nowhere. Did you mean https://github.com/OpenZeppelin/openzeppelin-sdk/tree/master/examples/lib-simple?

Reply:
It will once https://github.com/OpenZeppelin/openzeppelin-sdk/pull/1366 is merged ;-)